### PR TITLE
Loop Regions

### PR DESCRIPTION
**Pull request by phschaad**
This PR adds `LoopRegion`s to SDFGs. This forms the second core element of the [plan to make loops first class citizens of SDFGs](https://github.com/orgs/spcl/projects/10).
`LoopRegion`s are a special class of `ControlFlowRegion`s that represent different types of loops, meaning the control flow region inside them is executed a parametric number of times.
A `LoopRegion` _must_ have a conditional expression that determines whether the region is executed or not. It may additionally have:
- an initialization expression that is run when the region is first executed (even if the condition to execute its contents does not hold)
- an update expression, executed each time the execution of the region contents is finished
- a flag indicating that it's inverted, meaning the contents are executed once before the condition is checked (condition is checked again after every execution)
- a set of control flow blocks in its contents that, when reached, finish the execution of the region and do not execute it again even if the condition still holds (equivalent to `break` in C/Python)
- a set of control flow blocks that end the execution of its contents and execute the update statement if there is one. After this, the condition is checked again and the contents may be run again (equivalent to `continue` in C/Python)

For more general information on control flow regions please refer to [the documentation](https://spcldace.readthedocs.io/en/latest/sdfg/ir.html#elements) or the [PR that introduced them](https://github.com/spcl/dace/pull/1404).

An example of a tripple loop nest of regular for loops can be seen in the GEMM program below, showing that a proof-of-concept visualization for the introduced concepts is already available in the latest release version of the [VSCode extension](https://marketplace.visualstudio.com/items?itemName=phschaad.sdfv) (version 1.6.0 and upwards):

image

As outlined by the [project plan](https://github.com/orgs/spcl/projects/10), these `LoopRegion`s are currently _not_ being used by any frontend. They can, however, already be manually used through the SDFG builder API. According to plan, most passes and transformations are not currently able to handle the use of such `LoopRegion`s yet, so their use is still highly experimental. To allow traditional transformations and passes to work, as well as to be able to generate codes for SDFGs containing `LoopRegion`s, a compatibility pass (`InlineLoopRegions`) is available, together with a utility function `dace.sdfg.utils.inline_loop_blocks` that inlines any `LoopRegion`s to traditional state machine loops.

In summary, the PR includes:
- [x] Add `LoopRegion`s to SDFGs, a special control flow region to represent loops
- [x] Pass / Utility function to remove all loop regions from a graph, turning them back into regular state machine loops (a.k.a., inlining)
- [x] Tests for inlining of loops
- [x] Tests for the functional correctness of different types of loops (regular for, while, do-while, do-for) and nested loops (gemm - tripple nested loop)

### CODE REVIEW

**Review by alexnick83 — APPROVED** [review]
In general, looks good.

File: dace/sdfg/state.py [hunk 1]
```diff
@@ -618,7 +622,7 @@ def is_leaf_memlet(self, e):
 
     def used_symbols(self, all_symbols: bool, keep_defined_in_mapping: bool=False) -> Set[str]:
         state = self.graph if isinstance(self, SubgraphView) else self
-        sdfg = state.parent
+        sdfg = state.sdfg
```
[the remark this answers is not in the document]
Comment (phschaad) [reply dace/sdfg/state.py:625]: Correct

File: dace/sdfg/state.py [hunk 2]
```diff
@@ -2554,3 +2586,113 @@ def start_block(self, block_id):
             raise ValueError('Invalid state ID')
         self._start_block = block_id
         self._cached_start_block = self.node(block_id)
+
+
+@make_properties
+class LoopRegion(ControlFlowRegion):
+
```
[the remark this answers is not in the document]
Comment (phschaad) [reply dace/sdfg/state.py:2593]: Correct on the break / continue. I will make sure to add the necessary documentation, thanks!

File: dace/sdfg/state.py [hunk 3]
```diff
@@ -99,6 +99,10 @@ def in_degree(self, node: NodeT) -> int:
     def out_degree(self, node: NodeT) -> int:
         ...
 
+    @property
+    def sdfg(self) -> 'dace.sdfg.SDFG':
```
Comment (tbennun) [inline dace/sdfg/state.py:103]: you should probably import dace.sdfg.SDFG in the TYPE_CHECKING clause (line 28)

File: tests/sdfg/loop_region_test.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+    a_validation = np.zeros([10], dtype=np.float32)
+    a_test = np.zeros([10], dtype=np.float32)
+    sdfg(A=a_test)
+    for i in range(10):
+        a_validation[i] = i
+    assert np.allclose(a_validation, a_test)
+
+
+def test_tripple_nested_for():
```
Comment (tbennun) [inline tests/sdfg/loop_region_test.py:114]:
```suggestion
def test_triple_nested_for():
```

File: tests/sdfg/loop_region_test.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+    assert np.allclose(C_validation, C_test)
+
+
+if __name__ == '__main__':
+    test_loop_regular_for()
+    test_loop_regular_while()
+    test_loop_do_while()
+    test_loop_do_for()
+    test_tripple_nested_for()
```
Comment (tbennun) [inline tests/sdfg/loop_region_test.py:172]:
```suggestion
    test_triple_nested_for()
```

File: tests/transformations/control_flow_inline_test.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    init_edges = sdfg.out_edges(init_state)
+    assert len(init_edges) == 1
+    assert init_edges[0].data.assignments == {'i': '0'}
+    assert state0 in states
+    assert state1 in states
+    assert state2 in states
+    assert state3 in states
+
+
+def test_inline_tripple_nested_for():
```
Comment (tbennun) [inline tests/transformations/control_flow_inline_test.py:145]:
```suggestion
def test_inline_triple_nested_for():
```

File: tests/transformations/control_flow_inline_test.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    assert len(sdfg.edges_between(state4, end_state)) == 1
+    assert len(sdfg.edges_between(state2, tail_state)) == 1
+
+
+if __name__ == '__main__':
+    test_loop_inlining_regular_for()
+    test_loop_inlining_regular_while()
+    test_loop_inlining_do_while()
+    test_loop_inlining_do_for()
+    test_inline_tripple_nested_for()
```
Comment (tbennun) [inline tests/transformations/control_flow_inline_test.py:230]:
```suggestion
    test_inline_triple_nested_for()
```

File: dace/transformation/interstate/control_flow_inline.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        # Redirect all edges exiting the loop to instead exit the end state.
+        for a_edge in parent.out_edges(self.loop):
+            parent.add_edge(end_state, a_edge.dst, a_edge.data)
+            parent.remove_edge(a_edge)
+
+        # Add an initialization edge that initializes the loop variable if applicable.
+        init_edge = InterstateEdge()
+        if self.loop.init_statement is not None:
+            init_edge.assignments = {
+                self.loop.loop_variable: self.loop.init_statement.as_string.rpartition('=')[2].strip()
```
Comment (tbennun) [inline dace/transformation/interstate/control_flow_inline.py:75]: I am not super happy with this. There can be more than one statement in a codeblock, or a different statement entirely (i.e., that does not assign anything). Perhaps verify that there is only one ast.Assign in the `can_be_applied` method.

File: dace/transformation/interstate/control_flow_inline.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        if self.loop.inverted:
+            parent.add_edge(init_state, internal_start, init_edge)
+        else:
+            parent.add_edge(init_state, guard_state, init_edge)
+
+        # Connect the loop tail.
+        update_edge = InterstateEdge()
+        if self.loop.update_statement is not None:
+            update_edge.assignments = {
+                self.loop.loop_variable: self.loop.update_statement.as_string.rpartition('=')[2].strip()
```
Comment (tbennun) [inline dace/transformation/interstate/control_flow_inline.py:86]: same here

File: dace/sdfg/state.py [hunk 10]
```diff
@@ -2554,3 +2586,113 @@ def start_block(self, block_id):
...
+        if is_continue:
+            if is_break:
+                raise ValueError('Cannot set both is_continue and is_break')
+            self.continue_states.add(self.node_id(node))
+        if is_break:
+            if is_continue:
+                raise ValueError('Cannot set both is_continue and is_break')
+            self.break_states.add(self.node_id(node))
+
+    def add_state(self, label=None, is_start_block=False, is_continue=False, is_break=False, *,
```
Comment (tbennun) [inline dace/sdfg/state.py:2687]: a bit of code duplication here. shouldn't this just call add_node internally?

File: dace/sdfg/state.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (phschaad) [reply dace/sdfg/state.py:2613]: Added documentation

File: dace/sdfg/sdfg.py [hunk 11]
```diff
@@ -2170,6 +2171,10 @@ def compile(self, output_file=None, validate=True) -> 'CompiledSDFG':
             # if the codegen modifies the SDFG (thereby changing its hash)
             sdfg.build_folder = build_folder
 
+            # Convert any loop constructs with hierarchical loop regions into simple 1-level state machine loops.
+            # TODO (later): Adapt codegen to deal with hierarchical CFGs instead.
+            sdutils.inline_loop_blocks(sdfg)
```
[the remark this answers is not in the document]
Comment (phschaad) [reply dace/sdfg/sdfg.py:2176]: Moved.

File: dace/sdfg/state.py (hunk 3 above)
Comment (phschaad) [reply dace/sdfg/state.py]: Done

File: dace/transformation/interstate/control_flow_inline.py (hunk 8 above)
Comment (phschaad) [reply dace/transformation/interstate/control_flow_inline.py]: Fixed - I'm allowing any number of assignments (but nothing else!) since this can be expressed with a valid SDFG (isedges can have an arbitrary number of assignments).

File: dace/transformation/interstate/control_flow_inline.py (hunk 9 above)
Comment (phschaad) [reply dace/transformation/interstate/control_flow_inline.py]: same here

File: dace/sdfg/state.py (hunk 10 above)
Comment (phschaad) [reply dace/sdfg/state.py:2710]: fixed

**Review by tbennun — APPROVED** [review]
